### ISSUE TITLE
Experiment with using `tracing` instead of `log`

### ISSUE THREAD

**Issue by kevinaboos** [theseus-os/Theseus#1014]
## Proposed Benefits
The `log` crate is decent and easy to support, but it lacks some of the nicer features of more powerful logging systems like `tracing`. For example, it'd be nice to have an easier way to disable all log statements in a certain crate or module, or of a certain level within a given file, etc. The structured logging and span functionality all look attractive too.

Most importantly, this would help a user/developer more easily filter out the log statements that are of interest to them in a specific testing/debugging scenario. With `log` macros, we just have to generally enable all logging levels in all crates, and then tediously grep for keywords or file names.

## Considerations
The `tracing` crate [does support `no_std`](https://docs.rs/tracing/latest/tracing/#crate-feature-flags), but interestingly enough it requires `alloc` support on `no_std`. So that means we can use it in most crates, but not all crates (i.e., those that must run during early bootstrap before heap allocation is initialized).

## First steps
If anyone is interested in helping out, the best/easiest first step would likely be to play around with `tracing` as [a drop-in replacement for `log`](https://docs.rs/tracing/latest/tracing/index.html#for-log-users), which is easy enough because it offers the same basic macros.

**Comment by cocodery** [comment]
It seems like a lot of dirty work to do. but maybe not a difficult work?
Almost every crate use `log` crate for tracing, right?
Maybe it's a way to get familiar with `Theseus `source code.

**Comment by noryve-03** [comment]
Hello. Been playing around with tracing on my local directory. Is there anyone working on this issue @kevinaboos ?

**Comment by kevinaboos** [comment]
Nope, as far as I know it's just an idea. Feel free to go for it!

**Comment by noryve-03** [comment]
This issue is intimidating as the scale is huge. Any ideas on incremental improvements? I was thinking of first making a PR for the kernel, then another for applications.

**Comment by kevinaboos** [comment]
Definitely! I understand the scale is intimidating. I think a good first step is to look at only the `kernel/logger` crate, which acts as a logger backend and implements the required traits from `log`.

The `tracing` crate can supposedly work as a drop-in replacement for `log`, so you should only need to look at our one `logger` crate to make the relevant changes.

Note that there is no difference in how an application vs a kernel crate does logging in Theseus, they currently all go to the same log stream.

**Comment by noryve-03** [comment]
Thank you for the answer! But if I change the backend, won't it also be necessary to replace each call to the `log` crate with a call to the `tracing` crate? Maybe I should not be worried about that? Let me first finish my changes on the `kernel/logger` crate, then I will worry about that.

**Comment by kevinaboos** [comment]
oh true, you may need to do that. I'm not sure, part of the fun is finding out! :smiley: 

AFAIK, `tracing` is a drop-in replacement for `log` (based on what i've heard), but you may be correct in that it doesn't directly work with crates that use macros from the actual `log` crate itself.

**Comment by fabracht** [comment]
I've experimented with migrating Theseus from log to tracing. What I did was to migrate the task crate as a proof-of-concept.

  Key findings:

  - apparently, 96 kernel crates currently use log
  - 63 crates cannot use tracing (no alloc` support - early boot/memory management)
  - Migration is straightforward - tracing does work as a drop-in replacement for basic logging
  - The migrated code compiles cleanly with no warnings

  Benefits identified:

  - Better filtering capabilities (per-crate, per-module)
  - Structured logging with spans
  - More efficient when logs are disabled
  - Better debugging context tracking

I'd be willing to tackle some of this, but I would like your input @kevinaboos about what are the priorities here.
Thanks
